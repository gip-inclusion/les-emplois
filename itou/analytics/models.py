import uuid

from django.db import models
from django.utils import timezone

from itou.common_apps.address.departments import DEPARTMENTS
from itou.users.enums import UserKind


class DatumCode(models.TextChoices):
    # Employee record - Base
    EMPLOYEE_RECORD_COUNT = "ER-001", "FS totales"
    EMPLOYEE_RECORD_DELETED = "ER-002", "FS (probablement) supprimées"
    # Employee record - Lifecycle
    EMPLOYEE_RECORD_PROCESSED_AT_FIRST_EXCHANGE = "ER-101", "FS intégrées (0000) au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_AT_FIRST_EXCHANGE = "ER-102", "FS avec une erreur au premier retour"
    EMPLOYEE_RECORD_WITH_ERROR_3436_AT_FIRST_EXCHANGE = "ER-102-3436", "FS avec une erreur 3436 au premier retour"
    EMPLOYEE_RECORD_WITH_AT_LEAST_ONE_ERROR = "ER-103", "FS ayant eu au moins un retour en erreur"
    # Approval - Base
    APPROVAL_COUNT = "AP-001", "PASS IAE total"
    APPROVAL_CANCELLED = "AP-002", "PASS IAE annulés"
    # Approval - PE notification cycle
    APPROVAL_PE_NOTIFY_SUCCESS = "AP-101", "PASS IAE synchronisés avec succès avec pole emploi"
    APPROVAL_PE_NOTIFY_PENDING = "AP-102", "PASS IAE en attente de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_ERROR = "AP-103", "PASS IAE en erreur de synchronisation avec pole emploi"
    APPROVAL_PE_NOTIFY_READY = "AP-104", "PASS IAE prêts à être synchronisés avec pole emploi"
    # Users
    USER_COUNT = "US-001", "Nombre d'utilisateurs"
    USER_JOB_SEEKER_COUNT = "US-011", "Nombre de demandeurs d'emploi"
    USER_PRESCRIBER_COUNT = "US-012", "Nombre de prescripteurs"
    USER_EMPLOYER_COUNT = "US-013", "Nombre d'employeurs"
    USER_LABOR_INSPECTOR_COUNT = "US-014", "Nombre d'inspecteurs du travail"
    USER_ITOU_STAFF_COUNT = "US-015", "Nombre d'administrateurs"
    # API usage
    API_TOTAL_CALLS = "API-001", "API : total d'appels reçus"
    API_TOTAL_UV = "API-002", "API : total de visiteurs uniques"
    API_TOTAL_CALLS_CANDIDATS = "API-003", "API candidats : total d'appels reçus"
    API_TOTAL_UV_CANDIDATS = "API-004", "API candidats : total de visiteurs uniques"
    API_TOTAL_CALLS_GEIQ = "API-005", "API GEIQ : total d'appels reçus"
    API_TOTAL_UV_GEIQ = "API-006", "API GEIQ : total de visiteurs uniques"
    API_TOTAL_CALLS_ER = "API-007", "API FS : total d'appels reçus"
    API_TOTAL_UV_ER = "API-008", "API FS : total de visiteurs uniques"
    API_TOTAL_CALLS_MARCHE = "API-009", "API Le marché : total d'appels reçus"
    API_TOTAL_CALLS_SIAES = "API-010", "API siaes : total d'appels reçus"
    API_TOTAL_UV_SIAES = "API-011", "API siaes : total de visiteurs uniques (par adresse IP)"
    API_TOTAL_CALLS_STRUCTURES = "API-012", "API structures : total d'appels reçus"


class Datum(models.Model):
    """Store an aggregated `value` of the `code` data point for the specified `bucket`."""

    id = models.UUIDField(primary_key=True, default=uuid.uuid4)

    code = models.TextField(choices=DatumCode.choices)
    bucket = models.TextField()
    value = models.IntegerField()  # Integer offers the best balance between range, storage size, and performance

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "data"
        unique_together = ["code", "bucket"]
        indexes = [models.Index(fields=["measured_at", "code"])]


class StatsDashboardVisit(models.Model):
    dashboard_id = models.IntegerField(verbose_name="ID tableau de bord Metabase")
    dashboard_name = models.TextField(verbose_name="nom de la vue du tableau de bord")
    department = models.CharField(verbose_name="département", choices=DEPARTMENTS.items(), max_length=3, null=True)
    region = models.TextField(verbose_name="région", null=True)
    current_company_id = models.IntegerField(verbose_name="ID entreprise courante", null=True)
    current_prescriber_organization_id = models.IntegerField(
        verbose_name="ID organisation prescriptrice courante", null=True
    )
    current_institution_id = models.IntegerField(verbose_name="ID institution courante", null=True)
    user_kind = models.TextField(verbose_name="type d'utilisateur", choices=UserKind.choices)
    user_id = models.IntegerField(verbose_name="ID utilisateur")

    measured_at = models.DateTimeField(default=timezone.now)  # Not using auto_now_add=True to allow overrides

    class Meta:
        verbose_name_plural = "visite de tableau de bord"
