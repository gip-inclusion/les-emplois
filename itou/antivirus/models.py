from django.db import models
from django.utils.text import capfirst

from itou.files.models import File


class Scan(models.Model):
    file = models.OneToOneField(File, on_delete=models.CASCADE)
    clamav_signature = models.TextField()
    clamav_completed_at = models.DateTimeField(null=True, verbose_name="analyse ClamAV le")
    clamav_infected = models.BooleanField(null=True, verbose_name="fichier infect√© selon ClamAV")
    comment = models.TextField(blank=True, verbose_name="commentaire")

    class Meta:
        verbose_name = "analyse antivirus"
        verbose_name_plural = "analyses antivirus"

    def __str__(self):
        text = f"{capfirst(self._meta.verbose_name)} {self.file_id}"
        if self.clamav_infected:
            text = f"[VIRUS] {text}"
        return text
