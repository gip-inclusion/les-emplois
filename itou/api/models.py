import secrets

from django.db import models
from django.utils import timezone

from itou.companies.models import Company


def _generate_key():
    return secrets.token_urlsafe()


class CompanyToken(models.Model):
    key = models.CharField(default=_generate_key, unique=True, editable=False)
    label = models.CharField(verbose_name="m√©mo permettant d'identifier l'usage du jeton", max_length=60, unique=True)
    created_at = models.DateTimeField(default=timezone.now)
    companies = models.ManyToManyField(Company, related_name="api_tokens")

    class Meta:
        verbose_name = "jeton d'API entreprise"
        verbose_name_plural = "jetons d'API entreprise"
