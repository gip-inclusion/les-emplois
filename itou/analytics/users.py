from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
