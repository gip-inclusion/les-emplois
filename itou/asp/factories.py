import datetime
import random

import factory

from itou.asp import models


_sample_europe_countries = [
    {"code": "101", "name": "DANEMARK", "group": "2"},
    {"code": "111", "name": "BULGARIE", "group": "2"},
    {"code": "135", "name": "PAYS-BAS", "group": "2"},
]

_sample_outside_europe_countries = [
    {"code": "212", "name": "AFGHANISTAN", "group": "3"},
    {"code": "436", "name": "BAHAMAS", "group": "3"},
    {"code": "324", "name": "CONGO", "group": "3"},
]

_sample_france = [
    {"code": "714", "name": "BORA-BORA", "group": "1"},
    {"code": "812", "name": "KOUMAC", "group": "1"},
    {"code": "737", "name": "PUKAPUKA", "group": "1"},
]

_sample_communes = [
    {"code": "64483", "name": "SAINT-JEAN-DE-LUZ"},
    {"code": "97108", "name": "CAPESTERRE-DE-MARIE-GALANT"},
    {"code": "97107", "name": "CAPESTERRE-BELLE-EAU"},
    {"code": "37273", "name": "VILLE-AUX-DAMES"},
    {"code": "13200", "name": "MARSEILLE"},
    {"code": "67152", "name": "GEISPOLSHEIM"},
    {"code": "85146", "name": "MONTAIGU"},
]


class CountryFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = models.Country
        django_get_or_create = ("code",)

    code, name, group = random.choice(_sample_europe_countries).values()


class CountryFranceFactory(CountryFactory):
    code, name, group = "100", "FRANCE", "1"


class CountryCollectiviteOutremerFactory(CountryFactory):
    """
    Parts or regions of France having their own INSEE country code

    Most of them are "Collectivit√©s d'Outre-Mer"
    """

    code, name, group = random.choice(_sample_france).values()


class CountryEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_europe_countries).values()


class CountryOutsideEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_outside_europe_countries).values()


class CommuneFactory(factory.django.DjangoModelFactory):
    """Factory for ASP INSEE commune:
    - if `code` or `name` are build parameters, object values will result of a lookup in `_sample_communes`
    - otherwise, fields `code` and `name` will be set from a randomly picked sample commune
    """

    class Meta:
        model = models.Commune
        django_get_or_create = (
            "code",
            "start_date",
            "end_date",
        )

    # FIXME: may cause issues in testing validity periods
    start_date = datetime.date(2000, 1, 1)
    end_date = None

    # Skipping definition of `code` and `name` fields
    # will be set in `_adjust_kwargs`

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs


# FIXME: unreliable and confusing
class MockedCommuneFactory(CommuneFactory):
    """
    A factory with a specific code for mock testing
    """

    code, name = _sample_communes[-1].values()
