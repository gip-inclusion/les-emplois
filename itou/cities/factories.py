import json
import os

from django.contrib.gis.geos import GEOSGeometry, Point
from django.template.defaultfilters import slugify

from itou.cities.models import City


CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))

CITIES_JSON_FILE = f"{CURRENT_DIR}/management/commands/data/cities.json"


def create_test_cities(selected_departments, num_per_department=None):
    """
    Not a factory strictly speaking, but it's sitting here for discoverability.

    Pick cities from `cities.json` based on the given department numbers.
        create_test_cities(['01', '02'])

    The number of cities in each department can be limited:
        create_test_cities(['54', '57'], num_per_department=2)
    """
    departments_counter = {department: 0 for department in selected_departments}

    with open(CITIES_JSON_FILE, "r") as raw_json_data:

        json_data = json.load(raw_json_data)

        for item in json_data:

            department = item.get("codeDepartement")
            if (not department) or (department not in selected_departments):
                continue

            coords = item.get("centre")
            if not coords:
                continue

            departments_counter[department] += 1
            if num_per_department and departments_counter[department] > num_per_department:
                if all(value >= num_per_department for value in departments_counter.values()):
                    break
                continue

            City.objects.create(
                slug=slugify(f"{item['nom']}-{department}"),
                department=department,
                name=item["nom"],
                post_codes=item["codesPostaux"],
                code_insee=item["code"],
                coords=GEOSGeometry(f"{coords}"),  # Feed `GEOSGeometry` with GeoJSON.
            )


def create_city_saint_andre():
    return City.objects.create(
        name="Saint-André-des-Eaux",
        slug="saint-andre-des-eaux-44",
        department="44",
        coords=Point(-2.3140436, 47.3618584),
        post_codes=["44117"],
        code_insee="44117",
    )


def create_city_guerande():
    return City.objects.create(
        name="Guérande",
        slug="guerande-44",
        department="44",
        coords=Point(-2.4747713, 47.3358576),
        # Dummy
        post_codes=["44350"],
        code_insee="44350",
    )


def create_city_vannes():
    return City.objects.create(
        name="Vannes",
        slug="vannes-56",
        department="56",
        coords=Point(-2.8186843, 47.657641),
        # Dummy
        post_codes=["56000"],
        code_insee="56000",
    )
