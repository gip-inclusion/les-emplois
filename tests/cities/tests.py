from django.contrib.gis.geos import Point
from django.core import management
from django.core.management import call_command
from pytest_django.asserts import assertQuerySetEqual

from itou.cities.management.commands.sync_cities import get_next_insee_code
from itou.cities.models import City, EditionModeChoices
from tests.cities.factories import create_city_guerande, create_test_cities
from tests.companies.factories import JobDescriptionFactory
from tests.jobs.factories import create_test_romes_and_appellations
from tests.users.factories import JobSeekerFactory


def test_create_test_cities():
    create_test_cities(["62", "67", "93"], num_per_department=10)
    assert City.objects.count() == 30
    assert City.objects.filter(department="62").count() == 10
    assert City.objects.filter(department="67").count() == 10
    assert City.objects.filter(department="93").count() == 10


def test_sync_cities(settings, caplog, respx_mock):
    settings.API_GEO_BASE_URL = "https://geo.foo"
    respx_mock.get(
        "https://geo.foo/communes?fields=nom,code,codesPostaux,codeDepartement,codeRegion,centre&format=json"
    ).respond(
        200,
        json=[
            {
                "centre": {"coordinates": [4.9306, 46.1517], "type": "Point"},
                "code": "01001",
                "codeDepartement": "01",
                "codeRegion": "84",
                "codesPostaux": ["01400", "01234"],
                "nom": "L'Abergement-Clémenciat",
            },
            {
                "centre": {"coordinates": [5.4247, 46.0071], "type": "Point"},
                "code": "01002",
                "codeDepartement": "01",
                "codeRegion": "84",
                "codesPostaux": ["01640"],
                "nom": "L'Abergement-de-Varey",
            },
        ],
    )
    respx_mock.get(
        "https://geo.foo/communes?fields=nom,code,codesPostaux,codeDepartement,codeRegion,centre"
        "&format=json&type=arrondissement-municipal"
    ).respond(
        200,
        json=[
            {
                "centre": {"coordinates": [5.3828, 43.3002], "type": "Point"},
                "code": "13201",
                "codeDepartement": "13",
                "codeRegion": "93",
                "codesPostaux": ["13001"],
                "nom": "Marseille 1er Arrondissement",
            },
            {
                "centre": {"coordinates": [5.3496, 43.3225], "type": "Point"},
                "code": "13202",
                "codeDepartement": "13",
                "codeRegion": "93",
                "codesPostaux": ["13002"],
                "nom": "Marseille 2e Arrondissement",
            },
        ],
    )

    create_city_guerande()  # will be removed

    # will modify, since it's the same INSEE code
    City.objects.create(
        code_insee="01001",
        name="Nouveau Nom de Ville",
        slug="nouvelle-ville-01",
        department="01",
        coords=Point(-2.4747713, 47.3358576),
        post_codes=["01234"],
    )

    # workaround the forced save() that sets MANUAL mode up
    City.objects.update(edition_mode=EditionModeChoices.AUTO)

    management.call_command("sync_cities", wet_run=True)
    assert caplog.messages[:-1] == [
        (
            "HTTP Request: GET https://geo.foo/communes"
            '?fields=nom%2Ccode%2CcodesPostaux%2CcodeDepartement%2CcodeRegion%2Ccentre&format=json "HTTP/1.1 200 OK"'
        ),
        (
            "HTTP Request: GET https://geo.foo/communes?fields="
            "nom%2Ccode%2CcodesPostaux%2CcodeDepartement%2CcodeRegion%2Ccentre&format=json"
            '&type=arrondissement-municipal "HTTP/1.1 200 OK"'
        ),
        "count=1 label=City had the same key in collection and queryset",
        "\tCHANGED name=Nouveau Nom de Ville changed to value=L'Abergement-Clémenciat",
        "\tCHANGED post_codes=['01234'] changed to value=['01234', '01400']",
        "\tCHANGED coords=SRID=4326;POINT (-2.4747713 47.3358576) changed to value=SRID=4326;POINT (4.9306 46.1517)",
        "count=3 label=City added by collection",
        '\tADDED {"centre": {"coordinates": [5.4247, 46.0071], "type": "Point"}, '
        '"code": "01002", "codeDepartement": "01", "codeRegion": "84", '
        '"codesPostaux": ["01640"], "nom": "L\'Abergement-de-Varey"}',
        '\tADDED {"centre": {"coordinates": [5.3828, 43.3002], "type": "Point"}, '
        '"code": "13201", "codeDepartement": "13", "codeRegion": "93", '
        '"codesPostaux": ["13001"], "nom": "Marseille 1er"}',
        '\tADDED {"centre": {"coordinates": [5.3496, 43.3225], "type": "Point"}, '
        '"code": "13202", "codeDepartement": "13", "codeRegion": "93", '
        '"codesPostaux": ["13002"], "nom": "Marseille 2e"}',
        "count=1 label=City removed by collection",
        "\tREMOVED Guérande (44)",
        "Removed location from count=0 JobDescription due to city deletion",
        "Removed insee_city from count=0 Company due to city deletion",
        "Removed insee_city from count=0 PrescriberOrganization due to city deletion",
        "Removed insee_city from count=0 User due to city deletion",
        "Removed insee_city from count=0 Institution due to city deletion",
        "successfully deleted count=1 cities insee_codes=['44350']",
        "successfully updated count=1 cities",
        "successfully created count=3 new cities",
        "count=0 cities to replace",
    ]
    assert caplog.messages[-1].startswith(
        "Management command itou.cities.management.commands.sync_cities succeeded in"
    )

    # Introduce a "bogus" item in the regular (non arrondissement) cities:
    # - "new" city in respect to its non-registered INSEE code yet
    # - even though the slug is already registered in our DB
    # Check that it does not crash; this demonstrates an "INSEE code change".
    respx_mock.get(
        "https://geo.foo/communes?fields=nom,code,codesPostaux,codeDepartement,codeRegion,centre&format=json"
    ).respond(
        200,
        json=[
            {
                "centre": {"coordinates": [4.9306, 46.1517], "type": "Point"},
                "code": "01001",
                "codeDepartement": "01",
                "codeRegion": "84",
                "codesPostaux": ["01234", "01400"],  # changing the post codes order should not change
                "nom": "L'Abergement-Clémenciat",
            },
            {
                "centre": {"coordinates": [4.9306, 46.1517], "type": "Point"},
                "code": "01003",  # same new INSEE code
                "codeDepartement": "01",
                "codeRegion": "84",
                "codesPostaux": ["01400"],
                "nom": "L'Abergement-de-Varey",
            },
        ],
    )

    # Introduce a change in one of the arrondissements: the change is now permanent
    # and any new automatic sync update to it will be skipped.
    caplog.clear()
    marseille_1er = City.objects.get(slug="marseille-1er-13")
    marseille_1er.name = "Marssssssssseillle bébé"
    marseille_1er.save()

    management.call_command("sync_cities", wet_run=True)
    assert caplog.messages[:-1] == [
        (
            "HTTP Request: GET https://geo.foo/communes"
            '?fields=nom%2Ccode%2CcodesPostaux%2CcodeDepartement%2CcodeRegion%2Ccentre&format=json "HTTP/1.1 200 OK"'
        ),
        (
            "HTTP Request: GET https://geo.foo/communes?fields="
            "nom%2Ccode%2CcodesPostaux%2CcodeDepartement%2CcodeRegion%2Ccentre&format=json"
            '&type=arrondissement-municipal "HTTP/1.1 200 OK"'
        ),
        "count=3 label=City had the same key in collection and queryset",
        "skipping manually edited city=Marssssssssseillle bébé (13) from update",
        "count=1 label=City added by collection",
        '\tADDED {"centre": {"coordinates": [4.9306, 46.1517], "type": "Point"}, '
        '"code": "01003", "codeDepartement": "01", "codeRegion": "84", '
        '"codesPostaux": ["01400"], "nom": "L\'Abergement-de-Varey"}',
        "count=1 label=City removed by collection",
        "\tREMOVED L'Abergement-de-Varey (01)",
        "Removed location from count=0 JobDescription due to city deletion",
        "Removed insee_city from count=0 Company due to city deletion",
        "Removed insee_city from count=0 PrescriberOrganization due to city deletion",
        "Removed insee_city from count=0 User due to city deletion",
        "Removed insee_city from count=0 Institution due to city deletion",
        "successfully deleted count=1 cities insee_codes=['01002']",
        "successfully updated count=0 cities",  # no update to post codes
        "successfully created count=1 new cities",
        "count=0 cities to replace",
    ]
    assert caplog.messages[-1].startswith(
        "Management command itou.cities.management.commands.sync_cities succeeded in"
    )

    assertQuerySetEqual(
        City.objects.all().order_by("code_insee"),
        [
            (
                "L'Abergement-Clémenciat",
                "labergement-clemenciat-01",
                "01",
                ["01234", "01400"],
                "01001",
                "SRID=4326;POINT (4.9306 46.1517)",
                "AUTO",
            ),
            (
                "L'Abergement-de-Varey",
                "labergement-de-varey-01",
                "01",
                ["01400"],
                "01003",
                "SRID=4326;POINT (4.9306 46.1517)",
                "AUTO",
            ),
            (
                "Marssssssssseillle bébé",
                "marseille-1er-13",
                "13",
                ["13001"],
                "13201",
                "SRID=4326;POINT (5.3828 43.3002)",
                "MANUAL",
            ),
            (
                "Marseille 2e",
                "marseille-2e-13",
                "13",
                ["13002"],
                "13202",
                "SRID=4326;POINT (5.3496 43.3225)",
                "AUTO",
            ),
        ],
        transform=lambda city: (
            city.name,
            city.slug,
            city.department,
            city.post_codes,
            city.code_insee,
            city.coords,
            city.edition_mode,
        ),
    )


def test_sync_cities_with_refill(settings, caplog, respx_mock):
    settings.API_GEO_BASE_URL = "https://geo.foo"
    settings.API_INSEE_METADATA_URL = "https://insee.foo/metadata/"

    respx_mock.get(
        "https://geo.foo/communes?fields=nom,code,codesPostaux,codeDepartement,codeRegion,centre&format=json"
    ).respond(
        200,
        json=[
            {
                "centre": {"coordinates": [4.9306, 46.1517], "type": "Point"},
                "code": "01001",
                "codeDepartement": "01",
                "codeRegion": "84",
                "codesPostaux": ["01400", "01234"],
                "nom": "L'Abergement-Clémenciat",
            },
            {
                "centre": {"coordinates": [5.3496, 43.3225], "type": "Point"},
                "code": "19248",
                "codeDepartement": "19",
                "codeRegion": "75",
                "codesPostaux": ["19210"],
                "nom": "Les Trois-Saints",
            },
        ],
    )
    respx_mock.get(
        "https://geo.foo/communes?fields=nom,code,codesPostaux,codeDepartement,codeRegion,centre"
        "&format=json&type=arrondissement-municipal"
    ).respond(
        200,
        json=[
            {
                "centre": {"coordinates": [5.3828, 43.3002], "type": "Point"},
                "code": "13201",
                "codeDepartement": "13",
                "codeRegion": "93",
                "codesPostaux": ["13001"],
                "nom": "Marseille 1er Arrondissement",
            },
        ],
    )

    create_city_guerande()  # will be deleted - INSEE code 44350
    saint_martin_sepert = City.objects.create(
        name="Saint-Martin-Sepert",
        slug="saint-martin-sepert-19",
        department="19",
        # Dummy
        coords=Point(-2.4747713, 47.3358576),
        post_codes=["19210"],
        code_insee="19223",
    )
    # workaround the forced save() that sets MANUAL mode up
    City.objects.update(edition_mode=EditionModeChoices.AUTO)

    respx_mock.get("https://insee.foo/metadata/geo/commune/19223/suivants?date=2020-01-01").respond(
        200,
        json=[
            {
                "code": "19248",
                "uri": "http://id.insee.fr/geo/commune/4ee6ee69-c313-4110-8d62-9ac2de6cce73",
                "type": "Commune",
                "dateCreation": "2025-01-01",
                "intituleSansArticle": "Trois-Saints",
                "typeArticle": "4",
                "intitule": "Les Trois-Saints",
            }
        ],
    )
    # These objects will need to be "refilled"
    job_seeker_to_refill = JobSeekerFactory(insee_city=saint_martin_sepert)
    create_test_romes_and_appellations(["M1805"], appellations_per_rome=3)
    job_description_to_refill = JobDescriptionFactory(location=saint_martin_sepert)

    management.call_command("sync_cities", wet_run=True)
    new_city = City.objects.get(code_insee="19248")
    assert caplog.messages[:-1] == [
        (
            "HTTP Request: GET https://geo.foo/communes"
            '?fields=nom%2Ccode%2CcodesPostaux%2CcodeDepartement%2CcodeRegion%2Ccentre&format=json "HTTP/1.1 200 OK"'
        ),
        (
            "HTTP Request: GET https://geo.foo/communes?fields="
            "nom%2Ccode%2CcodesPostaux%2CcodeDepartement%2CcodeRegion%2Ccentre&format=json"
            '&type=arrondissement-municipal "HTTP/1.1 200 OK"'
        ),
        "count=0 label=City had the same key in collection and queryset",
        "count=3 label=City added by collection",
        (
            '\tADDED {"centre": {"coordinates": [4.9306, 46.1517], "type": "Point"}, '
            '"code": "01001", "codeDepartement": "01", "codeRegion": "84", '
            '"codesPostaux": ["01400", "01234"], "nom": "L\'Abergement-Clémenciat"}'
        ),
        (
            '\tADDED {"centre": {"coordinates": [5.3828, 43.3002], "type": "Point"}, '
            '"code": "13201", "codeDepartement": "13", "codeRegion": "93", '
            '"codesPostaux": ["13001"], "nom": "Marseille 1er"}'
        ),
        (
            '\tADDED {"centre": {"coordinates": [5.3496, 43.3225], "type": "Point"}, '
            '"code": "19248", "codeDepartement": "19", "codeRegion": "75", '
            '"codesPostaux": ["19210"], "nom": "Les Trois-Saints"}'
        ),
        "count=2 label=City removed by collection",
        "\tREMOVED Saint-Martin-Sepert (19)",
        "\tREMOVED Guérande (44)",
        "Removed location from count=1 JobDescription due to city deletion",
        "Removed insee_city from count=0 Company due to city deletion",
        "Removed insee_city from count=0 PrescriberOrganization due to city deletion",
        "Removed insee_city from count=1 User due to city deletion",
        "Removed insee_city from count=0 Institution due to city deletion",
        "successfully deleted count=2 cities insee_codes=['19223', '44350']",
        "successfully updated count=0 cities",  # no update to post codes
        "successfully created count=3 new cities",
        "count=1 cities to replace",
        'HTTP Request: GET https://insee.foo/metadata/geo/commune/19223/suivants?date=2020-01-01 "HTTP/1.1 200 OK"',
        "Found count=1 replacements",
        (
            f"Refilled JobDescription.location for pk={job_description_to_refill.pk} "
            f"to city={new_city.pk} (previous=19223)"
        ),
        "successfully refilled count=1 new cities for JobDescription",
        "successfully refilled count=0 new cities for Company",
        "successfully refilled count=0 new cities for PrescriberOrganization",
        f"Refilled User.insee_city for pk={job_seeker_to_refill.pk} to city={new_city.pk} (previous=19223)",
        "successfully refilled count=1 new cities for User",
        "successfully refilled count=0 new cities for Institution",
    ]
    assert caplog.messages[-1].startswith(
        "Management command itou.cities.management.commands.sync_cities succeeded in"
    )
    job_seeker_to_refill.refresh_from_db()
    assert job_seeker_to_refill.insee_city.code_insee == "19248"
    job_description_to_refill.refresh_from_db()
    assert job_description_to_refill.location.code_insee == "19248"


def test_resolve_insee_cities(caplog, snapshot):
    guerande = create_city_guerande()  # Guérande, 44350
    user = JobSeekerFactory(city="GUERAND", post_code="44350", geocoding_score=0.9)
    non_resolved_user_1 = JobSeekerFactory(city="Guérande", post_code="54350", geocoding_score=0.9)
    non_resolved_user_2 = JobSeekerFactory(city="ERAND", post_code="44350", geocoding_score=0.9)
    call_command("resolve_insee_cities", wet_run=True, mode="job_seekers")
    assert caplog.messages[:-1] == snapshot(name="first_pass")
    assert caplog.messages[-1].startswith(
        "Management command itou.cities.management.commands.resolve_insee_cities succeeded in "
    )

    user.refresh_from_db()
    assert user.insee_city == guerande
    non_resolved_user_1.refresh_from_db()
    non_resolved_user_1.geocoding_score = 0.0
    non_resolved_user_2.refresh_from_db()
    non_resolved_user_2.geocoding_score = 0.0

    caplog.clear()
    # no users selected: they either have a city or a low geocoding score.
    call_command("resolve_insee_cities", wet_run=True, mode="job_seekers")
    assert caplog.messages[:-1] == snapshot(name="second_pass")
    assert caplog.messages[-1].startswith(
        "Management command itou.cities.management.commands.resolve_insee_cities succeeded in "
    )


def test_get_next_insee_code(settings, respx_mock):
    settings.API_INSEE_METADATA_URL = "https://insee.foo/metadata/"

    # Test chained calls
    respx_mock.get("https://insee.foo/metadata/geo/commune/14011/suivants?date=1944-01-01").respond(
        200,
        json=[
            {
                "code": "14011",
                "uri": "http://id.insee.fr/geo/commune/9437c71f-c249-423c-a251-70533ebcd194",
                "type": "Commune",
                "dateCreation": "1973-01-01",
                "dateSuppression": "2017-01-01",
                "intituleSansArticle": "Anctoville",
                "typeArticle": "1",
                "intitule": "Anctoville",
            }
        ],
    )
    respx_mock.get("https://insee.foo/metadata/geo/commune/14011/suivants?date=1973-01-01").respond(
        200,
        json=[
            {
                "code": "14011",
                "uri": "http://id.insee.fr/geo/commune/1a3d5580-cdac-4e01-851c-e7e6b941f499",
                "type": "Commune",
                "dateCreation": "2017-01-01",
                "dateSuppression": "2024-04-24",
                "intituleSansArticle": "Aurseulles",
                "typeArticle": "1",
                "intitule": "Aurseulles",
            }
        ],
    )
    respx_mock.get("https://insee.foo/metadata/geo/commune/14011/suivants?date=2017-01-01").respond(
        200,
        json=[
            {
                "code": "14581",
                "uri": "http://id.insee.fr/geo/commune/ddd9f1bd-2090-4346-a2a2-8f79376ab37e",
                "type": "Commune",
                "dateCreation": "2024-04-24",
                "intituleSansArticle": "Aurseulles",
                "typeArticle": "1",
                "intitule": "Aurseulles",
            }
        ],
    )
    assert get_next_insee_code("14011", date="1944-01-01") == "14581"
    assert get_next_insee_code("14011", date="1973-01-01") == "14581"
    assert get_next_insee_code("14011", date="2017-01-01") == "14581"

    # Case with 2 children:
    respx_mock.get("https://insee.foo/metadata/geo/commune/60054/suivants?date=2020-01-01").respond(
        200,
        json=[
            {
                "code": "60054",
                "uri": "http://id.insee.fr/geo/commune/b1953db5-5d8d-421a-aa36-0550242be63b",
                "type": "Commune",
                "dateCreation": "2024-01-01",
                "intituleSansArticle": "Beaumont-les-Nonains",
                "typeArticle": "0",
                "intitule": "Beaumont-les-Nonains",
            },
            {
                "code": "60694",
                "uri": "http://id.insee.fr/geo/commune/f43ab45c-1603-461e-9191-6f1824d6031b",
                "type": "Commune",
                "dateCreation": "2024-01-01",
                "intituleSansArticle": "Hauts-Talican",
                "typeArticle": "4",
                "intitule": "Les Hauts-Talican",
            },
        ],
    )
    assert get_next_insee_code("60054", date="2020-01-01") is None
