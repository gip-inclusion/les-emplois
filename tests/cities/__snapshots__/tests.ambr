# serializer version: 1
# name: test_resolve_insee_cities[first_pass]
  list([
    "! failed to find matching city for item.city='ERAND' item.post_code='44350'",
    "! failed to find matching city for item.city='GuÃ©rande' item.post_code='54350'",
    '> count=1 JobSeeker updated. err_count=2 JobSeeker without a resolution.',
  ])
# ---
# name: test_resolve_insee_cities[second_pass]
  list([
    '> count=0 JobSeeker updated. err_count=0 JobSeeker without a resolution.',
  ])
# ---
